### PR TITLE
Prevent schema validation on nodes annotated @schema/type any=True

### PR DESCRIPTION
**Pull request by gcheadle-vmware**

### CODE REVIEW

File: pkg/schema/annotations.go [hunk 1]
```diff
@@ -20,36 +19,35 @@ const (
...
-	any          bool
-	inferredType yamlmeta.Type
-	itemPosition *filepos.Position
+	any  bool
+	item yamlmeta.Node
 }
 
 type NullableAnnotation struct {
-	providedValueType yamlmeta.Type
-	itemPosition      *filepos.Position
```
Comment (pivotaljohn) [inline pkg/schema/annotations.go:34]: 👍🏻  Yes: a `NullType` should wrap the contained value, but conceptually, what need would the `NullableAnnotation` have for it.

File: pkg/schema/annotations.go [hunk 2]
```diff
@@ -20,36 +19,35 @@ const (
...
+	item yamlmeta.Node
 }
 
 // DefaultAnnotation is a wrapper for a value provided via @schema/default annotation
 type DefaultAnnotation struct {
 	val interface{}
 }
 
-func NewTypeAnnotation(ann template.NodeAnnotation, inferredType yamlmeta.Type, pos *filepos.Position) (*TypeAnnotation, error) {
+// NewTypeAnnotation checks the key-word argument provided via @schema/type annotation, and returns wrapper for the annotated node.
```
Comment (pivotaljohn) [inline pkg/schema/annotations.go]:
```suggestion
// NewTypeAnnotation checks the keyword argument provided via @schema/type annotation, and returns wrapper for the annotated node.
```
nit: "keyword" is one word.

File: pkg/schema/annotations.go [hunk 3]
```diff
@@ -20,36 +19,35 @@ const (
...
+func NewTypeAnnotation(ann template.NodeAnnotation, node yamlmeta.Node) (*TypeAnnotation, error) {
 	if len(ann.Kwargs) == 0 {
 		return nil, schemaAssertionError{
-			position:    pos,
+			position:    node.GetPosition(),
```
Comment (pivotaljohn) [inline pkg/schema/annotations.go:43]: 🤔 💭  once annotations have their own position, perhaps there's no need to pass in the `node`?

File: pkg/schema/annotations.go [hunk 4]
```diff
@@ -170,26 +174,36 @@ func collectAnnotations(item yamlmeta.Node) ([]Annotation, error) {
...
+		}
+	}
+	return anns, nil
+}
+
+func processOptionalAnnotation(node yamlmeta.Node, optionalAnnotation template.AnnotationName, t yamlmeta.Type) (Annotation, error) {
+	nodeAnnotations := template.NewAnnotations(node)
+
+	if nodeAnnotations.Has(optionalAnnotation) {
+		ann, _ := nodeAnnotations[optionalAnnotation]
```
Comment (pivotaljohn) [inline pkg/schema/annotations.go]:
```suggestion
		ann := nodeAnnotations[optionalAnnotation]
```
nit: we're ignoring the `ok` return param, anyway; we can just not include it. 👍🏻  (it was there before, but why not clean it up)

File: pkg/schema/annotations.go [hunk 5]
```diff
@@ -170,26 +174,36 @@ func collectAnnotations(item yamlmeta.Node) ([]Annotation, error) {
...
 			return nil, err
 		}
+		if ann != nil {
+			anns = append(anns, ann)
+		}
+	}
+	return anns, nil
+}
+
+func processOptionalAnnotation(node yamlmeta.Node, optionalAnnotation template.AnnotationName, t yamlmeta.Type) (Annotation, error) {
```
Comment (pivotaljohn) [inline pkg/schema/annotations.go]:
```suggestion
func processOptionalAnnotation(node yamlmeta.Node, optionalAnnotation template.AnnotationName, effectiveType yamlmeta.Type) (Annotation, error) {
```
💭  "length of a variable 'ought' to be in proportion to the distance it is used"; this function is close to a "screen-full"; what role is this `yamlmeta.Type` playing here (answering that reveals intention)?

This propagates up and down the call stack, here:
- `NewDocumentType()` / `NewMapItemType()` / `NewArrayItemType()`
- `getValue()`
- `collectValueAnnotations()`
- `NewDefaultAnnotation()` (In a conversation we were having, I think I had said this was the "inferred type" ... and I was wrong. 😬 )

the actual name is (obviously) implementor's choice:
- `effectiveType`
- `typeOfValue`
- `actualType`
- ...

File: pkg/schema/type.go [hunk 6]
```diff
@@ -105,10 +104,7 @@ func (n NullType) GetDefaultValue() interface{} {
 }
 
 func (a AnyType) GetDefaultValue() interface{} {
-	if a.ValueType == nil {
-		return nil
-	}
-	return a.ValueType.GetDefaultValue() // delegate GetDefaultValue() functions will make defensive copy 👍
+	return a.defaultValue
```
Comment (pivotaljohn) [inline pkg/schema/type.go:110]: I think the onus is back on this function to provide a defensive copy.

File: pkg/schema/annotations.go [hunk 7]
```diff
@@ -20,36 +19,35 @@ const (
...
 
 type NullableAnnotation struct {
-	providedValueType yamlmeta.Type
-	itemPosition      *filepos.Position
+	item yamlmeta.Node
 }
 
 // DefaultAnnotation is a wrapper for a value provided via @schema/default annotation
 type DefaultAnnotation struct {
 	val interface{}
```
Comment (pivotaljohn) [inline pkg/schema/annotations.go:36]: 🤔 💭  Curious that `TypeAnnotation` and `NullableAnnotation` both have a reference to their host `yamlmeta.Node` but `DefaultAnnotation` does not...

File: pkg/schema/annotations.go [hunk 8]
```diff
@@ -20,36 +19,35 @@ const (
...
 }
 
-func NewTypeAnnotation(ann template.NodeAnnotation, inferredType yamlmeta.Type, pos *filepos.Position) (*TypeAnnotation, error) {
+// NewTypeAnnotation checks the key-word argument provided via @schema/type annotation, and returns wrapper for the annotated node.
+func NewTypeAnnotation(ann template.NodeAnnotation, node yamlmeta.Node) (*TypeAnnotation, error) {
```
Comment (pivotaljohn) [inline pkg/schema/annotations.go:40]:
```suggestion
func NewTypeAnnotation(ann template.NodeAnnotation, item yamlmeta.Node) (*TypeAnnotation, error) {
```
🤔  this variable plays the exact role as `TypeAnnotation.item`; so either they are _both_ `node` or `item`....?!

**Review by pivotaljohn — CHANGES_REQUESTED** [review]
There are a couple of _actual_ requested changes.

There are many comments that are "food for thought".  These I prefixed with 🤔 💭 .

File: pkg/schema/type.go (hunk 6 above)
Comment (gcheadle-vmware) [reply pkg/schema/type.go:110]: Deep copy for when interface is a Node, but scalar values are copied through interface type.go:110

**Comment by gcheadle-vmware** [comment]
@cari-lynn noticed `node.GetValues()[0]` in several new places, and we think we should change the way this is accessed to avoid nil pointer errs. This change was made because the interface `ValueHoldingNode` was deleted.
🤔 💭 ? @jtigger

**Comment by pivotaljohn** [comment]
Hey @gcheadle-vmware, the conflicts came from my merging #540. Feel free to grab me if you want a buddy to resolve.

**Comment by gcheadle-vmware** [comment]
@jtigger, conflicts resolved and ready to merge

**Review by pivotaljohn — APPROVED** [review]
LGTM! 🙏🏻

**Pull request merged**